modules/calib3d/src/chessboard.cpp

```diff
@@ -0,0 +1,3209 @@
+            rotate(angle,filtered_h,gray_image.size(),images[i]);
+            rotate(angle,filtered_v,gray_image.size(),images[i+num]);
+        }
+        cv::merge(images,rotated_images[scale_id]);
+
+        // calc feature map
+        calcFeatureMap(rotated_images[scale_id],feature_maps[scale_id]);
+
+        // filter feature map to improve impulse responses
+        if(parameters.filter)
+        {
+            cv::Mat high,low;
+            cv::blur(feature_maps[scale_id],low,cv::Size(scale_size,scale_size));
+            int scale2 = int((scale_size/6))*2+1;
+            cv::blur(feature_maps[scale_id],high,cv::Size(scale2,scale2));
+            feature_maps[scale_id] = high-0.8*low;
+        }
+    }
+}
+
+void FastX::detectImpl(const cv::Mat& image,std::vector<cv::KeyPoint>& keypoints,std::vector<cv::Mat> &feature_maps,const cv::Mat &mask)const
+{
+    std::vector<cv::Mat> rotated_images;
+    detectImpl(image,rotated_images,feature_maps,mask);
+    findKeyPoints(feature_maps,keypoints,mask);
+}
+
+void FastX::detectImpl(InputArray image, std::vector<KeyPoint>& keypoints, InputArray mask)const
+{
+    std::vector<cv::Mat> feature_maps;
+    detectImpl(image.getMat(),keypoints,feature_maps,mask.getMat());
+}
+
+void FastX::detectImpl(const Mat& src, std::vector<KeyPoint>& keypoints, const Mat& mask)const
+{
+    std::vector<cv::Mat> feature_maps;
+    detectImpl(src,keypoints,feature_maps,mask);
+}
+
+
+Ellipse::Ellipse():
+    angle(0),
+    cosf(0),
+    sinf(0)
+{
+}
+
+Ellipse::Ellipse(const cv::Point2f &_center, const cv::Size2f &_axes, float _angle):
+    center(_center),
+    axes(_axes),
+    angle(_angle),
+    cosf(cos(-_angle)),
+    sinf(sin(-_angle))
+{
+}
+
+Ellipse::Ellipse(const Ellipse &other)
+{
+    center = other.center;
+    axes= other.axes;
+    angle= other.angle;
+    cosf = other.cosf;
+    sinf = other.sinf;
+}
+
+const cv::Size2f &Ellipse::getAxes()const
+{
+    return axes;
+}
+
+cv::Point2f Ellipse::getCenter()const
+{
+    return center;
+}
+
+void Ellipse::draw(cv::InputOutputArray img,const cv::Scalar &color)const
+{
+    cv::Mat image = img.getMat();
+    cv::ellipse(image,center,axes,360-angle/M_PI*180,0,360,color);
+}
+
+bool Ellipse::contains(const cv::Point2f &pt)const
+{
+    cv::Point2f ptc = pt-center;
+    float x = cosf*ptc.x+sinf*ptc.y;
+    float y = -sinf*ptc.x+cosf*ptc.y;
+    if(x*x/(axes.width*axes.width)+y*y/(axes.height*axes.height) <= 1.0)
+        return true;
+    return false;
+}
+
+
+// returns false if the angle from the line pt1-pt2 to the line pt3-pt4 is negative
+static bool checkOrientation(const cv::Point2f &pt1,const cv::Point2f &pt2,
+        const cv::Point2f &pt3,const cv::Point2f &pt4)
+{
+    cv::Point3f p1(pt2.x-pt1.x,pt2.y-pt1.y,0);
+    cv::Point3f p2(pt4.x-pt3.x,pt4.y-pt3.y,0);
+    return p1.cross(p2).z > 0;
+}
+
+static bool sortKeyPoint(const cv::KeyPoint &pt1,const cv::KeyPoint &pt2)
+{
+    return pt1.response >= pt2.response;
```

Comment:
Equality should be dropped from here.
if `sortKeyPoint(a, b)` is true, then `sortKeyPoint(b, a)` must be a false (it is "less" comparison functions, not "less or equal").